Fix TLS, `io.async()` and package fetching on Windows

castholm opened the pull request:
Closes #25776

This fixes the HTTPS/TLS and package manager problems on Windows. Tested against the repro in the linked issue as well as by `zig build --fetch`ing https://github.com/castholm/zig-examples/tree/master/breakout with an empty global package cache. ([Déjà vu, anyone?](https://github.com/ziglang/zig/pull/24915))

Cheekiness aside, there were two separate issues unrelated to each other that regressed the package manager on Windows: TLS certificate lifetimes not being validated correctly and `io.async()` closure contexts not being correctly aligned, which have now hopefully been fixed.

Semi-important change: This PR redefines `Clock.real` such that implementations should always return timestamps relative to the POSIX/Unix epoch 1970-01-01 00:00:00+00:00. Previously, implementations were allowed to use implementation-specific epochs, which means that there's no way for the user to translate returned timestamps to actual calendar dates without digging into implementation details of any particular `Io` implementation and is obviously a problem if `Io` is supposed to be the "bring-your-own-OS" interface. Redefining it in this manner fixes this.

There are other ways to solve this, such as adding new vtable functions for returning the implementation-specific epoch, but in terms of complexity this redefinition is by far the simplest solution and only amounts to a simple 96-bit integer addition's worth of overhead for OSes like Windows that use non-POSIX/Unix epochs.

castholm committed:
io: Translate Windows `Clock.real` timestamps to the POSIX/Unix epoch

This fixes `std.http.Client` TLS certificate validation on Windows.

castholm committed:
io: Redefine `Clock.real` to return timestamps relative to the POSIX/Unix epoch

`Clock.real` being defined to return timestamps relative to an
implementation-specific epoch means that there's currently no way for
the user to translate returned timestamps to actual calendar dates
without digging into implementation details of any particular `Io`
implementation. Redefining it to return timestamps relative to
1970-01-01T00:00:00Z fixes this problem.

There are other ways to solve this, such as adding a new vtable function
for returning the implementation-specific epoch, but in terms of
complexity this redefinition is by far the simplest solution and only
amounts to a simple 96-bit integer addition's worth of overhead on OSes
like Windows that use non-POSIX/Unix epochs.

castholm committed:
io: Correctly align async closure contexts

This fixes package fetching on Windows.

Previously, `Async/GroupClosure` allocations were only aligned for the
closure struct type, which resulted in panics when `context_alignment`
(or `result_alignment` for that matter) had a greater alignment.

Techatrix committed:
std.Io: fix calls on functions that return an array type

Techatrix commented on lib/std/Io/Threaded.zig:
```diff
@@ -425,18 +426,57 @@ const AsyncClosure = struct {
+        errdefer comptime unreachable;
+
+        const actual_context_offset = context_alignment.forward(@intFromPtr(ac) + @sizeOf(AsyncClosure)) - @intFromPtr(ac);
+        const actual_result_offset = result_alignment.forward(actual_context_offset + context.len);
```
Here is a case I found where the alignment of the result data is incorrect:
```zig
const std = @import("std");

pub fn main() !void {
    var buffer: [1024]u8 align(64) = undefined;
    var fba: std.heap.FixedBufferAllocator = .init(buffer[1..]);

    var threaded: std.Io.Threaded = .init(fba.allocator());
    defer threaded.deinit();
    const io = threaded.io();

    var future = io.async(foo, .{});
    _ = future.await(io);
}

fn foo() Align64 {
    return .{ .data = 5 };
}

const Align64 = struct {
    data: u8 align(64),
};
```

```
zig run sample.zig 
thread 49096 panic: incorrect alignment
/home/techatrix/repos/zig/lib/std/Io.zig:1538:44: 0x1163858 in start (std.zig)
            const result_casted: *Result = @ptrCast(@alignCast(result));
                                           ^
/home/techatrix/repos/zig/lib/std/Io/Threaded.zig:406:16: 0x108cede in start (std.zig)
        ac.func(ac.contextPointer(), ac.resultPointer());
               ^
/home/techatrix/repos/zig/lib/std/Io/Threaded.zig:188:26: 0x10d9a26 in worker (std.zig)
            closure.start(closure);
                         ^
/home/techatrix/repos/zig/lib/std/Thread.zig:558:13: 0x10b2df0 in callFn__anon_16045 (std.zig)
            @call(.auto, f, args);
            ^
/home/techatrix/repos/zig/lib/std/Thread.zig:1534:30: 0x108eb60 in entryFn (std.zig)
                return callFn(f, self.fn_args);
                             ^
/home/techatrix/repos/zig/lib/std/os/linux/x86_64.zig:105:5: 0x10b2ea5 in clone (std.zig)
    asm volatile (
    ^
Aborted (core dumped)
```

I suspect that the issue is with `actual_result_offset` because it tries to align the size/offset instead of the memory address like it's being done in `actual_context_offset`. I haven't looked to much into it so this may because of something else.

castholm commented:
Added context/result alignment tests and fixed the remaining alignment bugs, and picked @Techatrix's array type return type fix from #25868. Should be good to go now.

andrewrk reviewed (approved):
Thank you for the fix and the tests.

Indeed, feels like deja vu. Clear sign that we have insufficient testing for package fetching, particularly on Windows.
